### ISSUE TITLE
Type-stability of `bijector(d::Distribution)` not complete

### ISSUE THREAD

**Issue by torfjelde** [TuringLang/Bijectors.jl#40]
I've made an effort to make things in the interface type-stable, but there's one `bijector(d::Distribution)` call that is currently not type-stable: `Truncated`.

For `bijector(d::Truncated)` we need to decide which bijectors to compose depending on whether or not it's only lower-bounded, upper-bounded or both.

Also, at the moment, some other univariate distributions (all the ones defined in union `TransformDistributions`) also have unstable bijector calls because they are also using the implementation as `Truncated`, e.g. `ArcSine`, but these we can just implement on a case-by-case basis to get type-stability in all cases except Truncated.

In practice, I don't think type-stability of `bijector(d::Distribution)` is that important, as one will usually construct the distributions at the very beginning of the computation. Even so, for the cases where this is not the case it would be nice to have type-stability.

Also, in the case of `Truncated{<:Distribution}` the union is of size 3 which can possibly be dealt with by union-splitting since default split-size is 4.

```julia
julia> for d in uni_dists
           c = @code_typed bijector(d)
           if (c.second isa Union) || (c.second == Any)
               println("-----------------")
               println(d, ": ", c.second)
               println("(", minimum(d), ", ", maximum(d), ")")
               println(bijector(d))
           end
       end
-----------------
Arcsine{Float64}(a=2.0, b=4.0): Union{Identity, Logit{Float64}, Composed}
(2.0, 4.0)
Bijectors.Logit{Float64}(2.0, 4.0)
-----------------
Biweight{Float64}(μ=0.0, σ=1.0): Union{Identity, Logit{Float64}, Composed}
(-1.0, 1.0)
Bijectors.Logit{Float64}(-1.0, 1.0)
-----------------
Cosine{Float64}(μ=0.0, σ=1.0): Union{Identity, Logit{Float64}, Composed}
(-1.0, 1.0)
Bijectors.Logit{Float64}(-1.0, 1.0)
-----------------
Epanechnikov{Float64}(μ=0.0, σ=1.0): Union{Identity, Logit{Float64}, Composed}
(-1.0, 1.0)
Bijectors.Logit{Float64}(-1.0, 1.0)
-----------------
Levy{Float64}(μ=0.0, σ=1.0): Union{Identity, Composed{Tuple{Shift{Float64},Log}}}
(0.0, Inf)
Composed{Tuple{Bijectors.Shift{Float64},Bijectors.Log}}((Bijectors.Shift{Float64}(-0.0), Bijectors.Log()))
-----------------
Pareto{Float64}(α=1.0, θ=1.0): Union{Identity, Composed{Tuple{Shift{Float64},Log}}}
(1.0, Inf)
Composed{Tuple{Bijectors.Shift{Float64},Bijectors.Log}}((Bijectors.Shift{Float64}(-1.0), Bijectors.Log()))
-----------------
Truncated(Normal{Float64}(μ=0.0, σ=1.0), range=(-Inf, 2.0)): Union{Identity, Logit{Float64}, Composed}
(-Inf, 2.0)
Composed{Tuple{Bijectors.Scale{Float64},Composed{Tuple{Bijectors.Shift{Float64},Bijectors.Log}}}}((Bijectors.Scale{Float64}(-1.0), Composed{Tuple{Bijectors.Shift{Float64},Bijectors.Log}}((Bijectors.Shift{Float64}(2.0), Bijectors.Log()))))
```

**Comment by torfjelde** [comment]
I just want to point out that this has been solved for all cases except `Truncated` for which runtime information is required to be able to determine which transformation is necessary. This was done in PR #39.

**Comment by mohamed82008** [comment]
Can we define a generalization of the bijectors used for lower bounded, upper bounded and both? Would a coupling flow be useful here?

**Comment by torfjelde** [comment]
> Can we define a generalization of the bijectors used for lower bounded, upper bounded and both? Would a coupling flow be useful here?

Yeah, so that's one approach that came up at some point where we basically define a `TrunactedBijector` which just reimplements the `link` and `invlink` for these distributions. You could do something like 
```julia
# Could also implement this for the case of `Bijector{1}`
struct TruncatedBijector{T} <: Bijector{0}
    lb::T
    ub::T
end

function (b::TruncatedBijector)(x::Real)
    a, b = b.lb, b.ub
    lowerbounded, upperbounded = isfinite(a), isfinite(b)
    if lowerbounded && upperbounded
        return StatsFuns.logit((x - a) / (b - a))
    elseif lowerbounded
        return log(x - a)
    elseif upperbounded
        return log(b - x)
    else
        return x
    end 
end

function (ib::Inversed{<:TruncatedBijector})(y::Real)
    a, b = b.lb, b.ub
    lowerbounded, upperbounded = isfinite(a), isfinite(b)
    if lowerbounded && upperbounded
        return (b - a) * StatsFuns.logistic(y) + a
    elseif lowerbounded
        return exp(y) + a
    elseif upperbounded
        return b - exp(y)
    else
        return y
    end
end

function logabsdetjac(b::TruncatedBijector, x::Real)
    a, b = b.lb, b.ub
    lowerbounded, upperbounded = isfinite(a), isfinite(b)
    if lowerbounded && upperbounded
        return - log((x - a) * (b - x) / (b - a))
    elseif lowerbounded
        return - log(x - a)
    elseif upperbounded
        return - log(b - x)
    else
        return zero(x)
    end
end
```

**Comment by torfjelde** [comment]
Closed by #52
